File: php/src/MemoryRateLimiterStore.php
<?php

declare(strict_types=1);

namespace SecureKit;

/**
 * Process-local, fixed-window rate limiter store. Does NOT share state
 * across processes/instances — see THREAT_MODEL.md.
 */
final class MemoryRateLimiterStore implements RateLimiterStoreInterface
{
    /** @var array<string, array{count: int, resetAt: float}> */
    private array $buckets = [];

    public function increment(string $key, int $windowSeconds): int
    {
        $now = microtime(true);
        if (!isset($this->buckets[$key]) || $now > $this->buckets[$key]['resetAt']) {
            $this->buckets[$key] = ['count' => 0, 'resetAt' => $now + $windowSeconds];
        }
        $this->buckets[$key]['count']++;
        return $this->buckets[$key]['count'];
    }
}

File: java/src/main/java/io/github/securekit/MemoryRateLimiterStore.java
package io.github.securekit;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Process-local, thread-safe fixed-window rate limiter store. Does NOT
 * share state across processes/instances -- see THREAT_MODEL.md.
 */
public final class MemoryRateLimiterStore implements RateLimiterStore {

    private record Bucket(int count, long resetAt) {
    }

    private final ConcurrentHashMap<String, AtomicReference<Bucket>> buckets = new ConcurrentHashMap<>();

    @Override
    public int increment(String key, long windowMillis) {
        AtomicReference<Bucket> ref = buckets.computeIfAbsent(key, k -> new AtomicReference<>(new Bucket(0, 0L)));
        while (true) {
            Bucket current = ref.get();
            long now = System.currentTimeMillis();
            Bucket next = now > current.resetAt()
                    ? new Bucket(1, now + windowMillis)
                    : new Bucket(current.count() + 1, current.resetAt());
            if (ref.compareAndSet(current, next)) {
                return next.count();
            }
        }
    }
}
